### PR TITLE
Fix monomorphization and add tests

### LINKED ISSUE #624: Monomorphizer fails on over-application
```
(coalton-toplevel
  (define (f x y)
    (+ x y))

  (declare g (Num :a => :a -> :a))
  (define g
    (f 2))

  (monomorphize)
  (declare h (Integer -> Integer))
  (define (h y)
    (g y)))
```
Errors with:
`The bounding indices 2 and COMMON-LISP:NIL are bad for a sequence of length 1.`

### PR DESCRIPTION
**Pull request by eliaslfox**
Fixes #624

### CODE REVIEW

**Comment by stylewarning** [comment]
this front end refactor is paying dividends

File: src/utilities.lisp [hunk 1]
```diff
@@ -156,6 +157,18 @@
          (nreverse out)
          result)))))
 
+(defun take (n list)
```
Comment (stylewarning) [inline src/utilities.lisp:160]: these functions work for any sequence

**Review by colescott — APPROVED** [review]

File: src/codegen/monomorphize.lisp [hunk 2]
```diff
@@ -199,103 +199,104 @@ recompilation, and also maintains a stack of uncompiled candidates."
...
+
+                      (remaing-names (loop :for arg :in remaining-args
+                                           :collect (gentemp)))
+
+                      (remaining-types (loop :for arg :in remaining-args
+                                             :for ty :in (tc:function-type-arguments new-type)
+                                             :collect ty)))
+
+                 ;; All remaining arguments must be unpropagated
+                 (unless (every (lambda (x) (eq x '@@unpropagated)) remaining-args)
```
Comment (stylewarning) [inline src/codegen/monomorphize.lisp:260]: is it ok that EVERY is true for an empty list?

**Review by stylewarning — APPROVED** [review]

File: src/codegen/monomorphize.lisp (hunk 2 above)
Comment (eliaslfox) [reply src/codegen/monomorphize.lisp:260]: Because of the check in the cond branch the list will never be empty.

Comment (gefjon) [reply src/codegen/monomorphize.lisp:260]: This seems like a great thing to note in a quick comment.

File: src/utilities.lisp (hunk 1 above)
Comment (gefjon) [reply src/utilities.lisp:160]: Restricting their type with a declaration in a local function allows better type inference and checking, plus potentially compiling away some dispatch.

Comment (stylewarning) [reply src/utilities.lisp:160]: what you say is true yet i don't see many calls to LIST-LENGTH in our code

Comment (stylewarning) [reply src/utilities.lisp:160]: 😬

Beware, it returns NIL on circular lists.

**Pull request merged**